### DISCUSSION TITLE
sign verification function failed

### DISCUSSION THREAD

**Discussion by angieyth** [aptos-labs/aptos-developer-discussions#262] [Questions] [answered]
### Discord user ID

hazard

### Describe your question in detail.

So two months back I have deployed a NFT marketplace contract with sign verification function to purchase NFT which was working alright. Today after two months I deployed a fresh new contract then tried to purchase a listing but it gives me an error message of sign invalid. At the same time when I try to purchase a listing on old NFT marketplace contract its working fine. Am I doing something wrong? or should I update any dependencies, please let me know.

note: new contract code is exactly same as old one

old contract : https://explorer.aptoslabs.com/account/0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66/transactions?network=testnet

new contract : https://explorer.aptoslabs.com/account/0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e/transactions?network=testnet

### What error, if any, are you getting?

INVALID SIGN (an error defined in the contract) 

### What have you tried or looked at?  Or how can we reproduce the error?

N/A

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by 0x-j** [comment]
I think it's because you deployed new contract using a different address. 
New contract is deployed by `0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e`.
Old contract is deployed by `0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66`.
Can you try calling purchasing of old contract with signature of `0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66`?

**Reply by pravinho** [reply]
Actually both contracts are deployed using different address. For contract ```0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66``` I have deployed it like 2 months back and signature of ```0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66``` worked alright at that time and also its working fine today if I list and purchase a NFT. On the other hand contract ```0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e``` is a new contract I have deployed yesterday with a exact same code as one in ```0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66``` but it fails with an error ```Move abort in 0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e::nft_listing: EINVALID_SIGN(0x50003): The sign is invalid```. Script I use to generate signature are also same with the little changes as required.

**Reply by 0x-j** [reply]
Can you check the signature generation script? Im guessing it needs to generate signuater using the contract deployer, since you deploy with a different account, you need to generate signature using that new account

**Reply by pravinho** [reply]
1. If there is some issue with script then it should not be working on old contract but that is not a case here.
2. I am using same script for sign generation in old and new contract. Which work perfectly fine for old contract I deployed like 2 months ago but it gives out error in new contract which I deployed 2 days before today.
3. I am also aware that signing should be done by authorised signer we mentioned in contract. Which I did without mistake, is the reason why sign verification is going through in old contract. 

For one whole day I was trying sign verification in new contract it was not working but It worked in first attempt on old contract 

You can also check that both the old and new contract (nft_listing) code is exactly same apart from some comments removed in new contract

**Reply by Alexzander-Stone** [reply]
Hey there @pravinho. Looking at your [old contract](https://explorer.aptoslabs.com/account/0xb8ad2cf1e41d860890664fdd06a3c4f064fa246af47334690ea785e1407d7a66?network=testnet), you don't have a TX that invokes the `nft_listing::purchase` function. Was there a different contract or TX that showcases this working?

The error you're facing is one of your custom errors, `EINVALID_SIGN`. Following the path of what code should be called for `nft_listing::purchase`, this error happens during the `signVerification` function call. Which implies that either your supplied signature data is incorrect, or the signature verification process is incorrectly defined.

**Reply by pravinho** [reply]
Hi @Alexzander-Stone , 

This is successful `nft_listing::purchase`  on old contract's transaction link : https://explorer.aptoslabs.com/txn/0x93a488d3cac6294c3d151abd0b615b220b38a7a534379358095cfc24861ce749/userTxnOverview?network=testnet

Well below code is what is use for sign generation which works fine in old contract :  

```import {
    Account,
    AccountAddress,
    AnyNumber,
    Aptos,
    AptosConfig,
    MoveString,
    Network,
    NetworkToNetworkName,
    Serializable,
    Serializer,
    U64,
    Ed25519PrivateKey,
    Hex,
    AccountAddressInput,
    HexInput,
  } from "@aptos-labs/ts-sdk";
  
  const config = new AptosConfig({ network: Network.TESTNET });
  const aptos = new Aptos(config);
  
  export class MessageData extends Serializable {
    // Contract's address
    public readonly contractAddress: AccountAddress;
  
    // Module name, i.e: 0x1::account
    public readonly moduleName: MoveString = new MoveString("nft_listing");
  
    // The struct name
    public readonly structName: MoveString = new MoveString("MesssageData");
  
    public readonly listing_addr: AccountAddress;
  
    public readonly purchaser_addr: AccountAddress;
  
    public readonly sequence_number: U64;
  
    public readonly salt: U64;
  
    constructor(args: {
      contractAddress: AccountAddress;
      listing_addr: AccountAddressInput;
      purchaser_addr: AccountAddressInput;
      sequence_number: AnyNumber;
      salt: AnyNumber;
    }) {
      super();
      this.contractAddress = args.contractAddress;
      this.listing_addr = AccountAddress.from(args.listing_addr);
      this.purchaser_addr = AccountAddress.from(args.purchaser_addr);
      this.sequence_number = new U64(args.sequence_number);
      this.salt = new U64(args.salt);
    }
  
    serialize(serializer: Serializer): void {
      serializer.serialize(this.contractAddress);
      serializer.serialize(this.moduleName);
      serializer.serialize(this.structName);
      serializer.serialize(this.listing_addr);
      serializer.serialize(this.purchaser_addr);
      serializer.serialize(this.sequence_number);
      serializer.serialize(this.salt);
    }
  }
  
  async function main() {
  // Create a private key instance for Ed25519 scheme
  const privateKey = new Ed25519PrivateKey("");

  const account = await Account.fromPrivateKey({ privateKey });
  
  const listing_addr = "0x6daa139e2553916785749912feb780e2e6e64a3a49fdfc7fff6c2dbd4cc18a7a";
  const purchaser_addr = "0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e";
  const sequence_number = 3;
  const salt = 12345;
  
  
  // console.log(contractAddressu8)
  
  
  const msg = new MessageData({
    contractAddress: account.accountAddress,
    listing_addr : listing_addr,
    purchaser_addr: purchaser_addr,
    sequence_number: sequence_number,
    salt: salt,
  });
  
  const serializer = new Serializer();
  serializer.serialize(msg);
  const signature = account.sign(serializer.toUint8Array());
  
  console.log("Signature ==========> ",signature)
  
  }
  
  main()
  
  ```

**Reply by Alexzander-Stone** [reply]
Hi @pravinho, after looking through execution more, I believe the signed message must be at fault.
Object listing, function-id, and type args are all correct. Which leaves the signed message and the salt as possible fault locations.


Could you double check that you're defining your MessageData variables appropriately? And then post the values you used, below?
Specifically:
* `listing_addr`
* `purchaser_addr`
* `sequence_number`
* `salt`

Additionally, double check your `privateKey` that is used to create the account (should be related to the contract address of: `0x86c5076a4512ed215b457b1c8f8e5214bb1a5fb25abce3f8eb8713f93cd98e5e`). I'd recommend performing a quick log statement on the `account.accountAddress`, and sharing that below, too.

**Reply by pravinho** [reply, marked as answer]
hey @Alexzander-Stone, 

The @ownerPublicKey input set on Move.toml which is given in during deployment was not right. Which led to the sign validation error. Thank you for your support, issue has been resolved now.
